### PR TITLE
Exclusive shape yyc

### PR DESCRIPTION
**Pull request by yyc84**
exclusive shape template

### CODE REVIEW

File: src/shared/geometries/exclusive_shape.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        };
+
+        virtual Vecd findClosestPoint(const Vecd &probe_point) override
+        {
+            return base_shape_->findClosestPoint(probe_point);
+        };
+
+      private:
+            UniquePtrKeeper<BaseShapeType> base_shape_keeper_;
+            Shape *base_shape_;
```
Comment (Xiangyu-Hu) [inline src/shared/geometries/exclusive_shape.h:71]: you do not need these?

File: src/shared/geometries/exclusive_shape.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            base_shape_(base_shape_keeper_.template createPtr<BaseShapeType>(std::forward<ConstructorArgs>(args)...))
+        {};
+
+        virtual ~ExclusiveShape(){};
+
+        /*reverse the value of checkContain function*/
+        virtual bool checkContain(const Vecd &probe_point, bool BOUNDARY_INCLUDED = true) override
+        {
+            bool is_inside = base_shape_->checkContain(probe_point);
+            return !is_inside;
```
Comment (Xiangyu-Hu) [inline src/shared/geometries/exclusive_shape.h:61]: simply use:
return !BaseShapeType::checkContain(probe_point);

File: src/shared/geometries/exclusive_shape.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        };
+
+        virtual Vecd findClosestPoint(const Vecd &probe_point) override
+        {
+            return base_shape_->findClosestPoint(probe_point);
+        };
```
Comment (Xiangyu-Hu) [inline src/shared/geometries/exclusive_shape.h:67]: you do not need this.

**Review by Xiangyu-Hu — CHANGES_REQUESTED** [review]

File: src/shared/geometries/exclusive_shape.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+#define EXCLUSIVE_SHAPE_H
+
+#include "base_data_package.h"
+#include "base_geometry.h"
+
+namespace SPH
+{
+/**
+ * @class ExclusiveShape
+ * @brief A template shape which has the fluid outside of the geometry.
```
[the remark this answers is not in the document]
Comment (ChiZhangatTUM) [reply src/shared/geometries/exclusive_shape.h:39]: If I understood this correctly, this should be kind of masking scheme with level-set method. Can we name it MaskingGeometry or MaskingSahpe?

File: src/shared/geometries/exclusive_shape.h (hunk 2 above)
Comment (yyc84) [reply src/shared/geometries/exclusive_shape.h:61]: ok

File: src/shared/geometries/exclusive_shape.h (hunk 3 above)
Comment (yyc84) [reply src/shared/geometries/exclusive_shape.h:67]: ok

File: src/shared/geometries/exclusive_shape.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (yyc84) [reply src/shared/geometries/exclusive_shape.h:39]: region outside shape seems better

[the remark this answers is not in the document]
Comment (yyc84) [reply src/shared/geometries/exclusive_shape.h:39]: it is not only for level-set, it give a negative distance value of orginal shape

[the remark this answers is not in the document]
Comment (ChiZhangatTUM) [reply src/shared/geometries/exclusive_shape.h]: The name should clearly presents the purpose of this specific region. For example ExclusiveShape indicating that the shape is excluded, while not clearly indicates for which purpose is excluded.

[the remark this answers is not in the document]
Comment (Xiangyu-Hu) [reply src/shared/geometries/exclusive_shape.h]: Probably using InverseShape?

[the remark this answers is not in the document]
Comment (ChiZhangatTUM) [reply src/shared/geometries/exclusive_shape.h]: > Probably using InverseShape?

This one is better than ExclusiveShape

**Comment by yyc84** [comment]
I passed windows on my own computer, and the case "3d_stfb" has no problem on windows computer

**Review by Xiangyu-Hu — APPROVED** [review]

**Pull request merged**